### ISSUE TITLE
Decay-based purging not working if a thread calls free() before any *alloc() call

### ISSUE THREAD

**Issue by kempniu** [jemalloc/jemalloc#2251]
First of all, thank you for your work on jemalloc!

While investigating a case of an application's VSZ and RSS growing indefinitely over time on FreeBSD 12.3 (which uses jemalloc 5.1.0), I noticed a peculiar jemalloc behavior which seems to be the root cause.

***NOTE:** This problem seems to have been addressed in the `dev` branch, but not in any numbered jemalloc release and I wonder whether it is a known issue.*

If the first allocator-related action for a given thread is a `free()`, then:

 1. A minimally-initialized tsd is [set up][1].
 2. This causes the `arenas_tdata_bypassp` flag to be [set][2] to `true`.
 3. The `arenas_tdata_bypassp` flag remains set to `true` even after the tsd transitions to the nominal state (e.g. due to a `malloc()` call).  (I could not find any code path that would set that flag to `false` after it is set to `true` in `tsd_data_init_nocleanup()`.)
 4. As long as the `arenas_tdata_bypassp` flag is set to `true`, [`arena_tdata_get_hard()` will not allocate `arenas_tdata`][3], causing that function to always return `NULL` for that thread.
 5. This causes `arena_tdata_get()` to [always return `NULL` for that thread][4].
 6. This causes `decay_ticker_get()` to [always return `NULL` for that thread][5].
 7. This [prevents ticker-based decay from ever being triggered for that thread][6].

If an application's allocation/deallocation patterns repeatedly cause dirty pages to be created, this can lead to memory exhaustion over time.  Commit fb56766ca9b398d07e2def5ead75a021fc08da03 (5.2.0+) helps in my case because it adds another way for large extents to be purged, but AFAICT it does not completely solve the issue; that only happened in c259323ab3082324100c708109dbfff660d0f4b8, which was not yet included in any numbered jemalloc release (it is only present on the `dev` branch).

I confirmed that making each application thread issue a `malloc()` call right at the beginning of the thread function enables decay-based purging to work in jemalloc versions 5.0.0 through 5.2.1.

Is this a known issue with jemalloc 5.0.0+ versions released to date?  I could not find the rationale behind commit c259323ab3082324100c708109dbfff660d0f4b8 anywhere; the code comment it added says that the intention is "to avoid having per-arena state in the tsd", but I do not know whether that was done purely as a performance improvement/optimization or as a bug fix as well.

As a side note, does c259323ab3082324100c708109dbfff660d0f4b8 not make #684 irrelevant?

Thank you in advance for sharing any thoughts you might have on the above!

[1]: https://github.com/jemalloc/jemalloc/blob/5.1.0/src/jemalloc.c#L2377-L2385
[2]: https://github.com/jemalloc/jemalloc/blob/5.1.0/src/tsd.c#L108
[3]: https://github.com/jemalloc/jemalloc/blob/5.1.0/src/jemalloc.c#L436-L445
[4]: https://github.com/jemalloc/jemalloc/blob/5.1.0/include/jemalloc/internal/jemalloc_internal_inlines_a.h#L60-L65
[5]: https://github.com/jemalloc/jemalloc/blob/5.1.0/include/jemalloc/internal/jemalloc_internal_inlines_a.h#L102-L105
[6]: https://github.com/jemalloc/jemalloc/blob/5.1.0/include/jemalloc/internal/arena_inlines_b.h#L69-L72

**Comment by interwq** [comment]
Thanks for the thorough investigation and detailed report! I wasn't aware of this issue. It looks to me that `arenas_tdata_bypassp` indeed wasn't set up correctly, when a thread does a `free()` first in its lifetime.

From what I remember, c259323 was motivated by avoiding cache / TLB misses. It happens to resolve this issues also shows the benefit of the simplified implementation.

> As a side note, does https://github.com/jemalloc/jemalloc/commit/c259323ab3082324100c708109dbfff660d0f4b8 not make https://github.com/jemalloc/jemalloc/issues/684 irrelevant?

It does. Just closed #684.

We are really close to releasing 5.3, with the [current dev branch](https://github.com/jemalloc/jemalloc/commit/ed5fc14b28ca62a6ba57b65adf557e1ef09037f0) being the release candidate. You can follow the [progress here](https://github.com/jemalloc/jemalloc/issues/2213). Afterwards we likely want to find some help integrating 5.3 into FreeBSD to resolve this issue.

Let's keep this issue open, as least until the new jemalloc gets integrated into FreeBSD, in case other people run into it as well.

**Comment by interwq** [comment]
@kempniu : forgot to mention, another option is to enable background threads (adding `background_thread:true` in `MALLOC_CONF`). It allows jemalloc to create its own threads which handles the purging and takes that load off the application threads. jemalloc can't really enable it by default but in the vast majority of use cases, it improves tail latency and smooths out the purging behavior (e.g. avoids RSS spikes when application threads go idle).

**Comment by kempniu** [comment]
Right, thanks.  A more radical workaround that also worked in our case was adding `dirty_decay_ms:0` to `MALLOC_CONF`.  While this change is obviously not universally feasible, it comes with a nice side effect of keeping RSS close to the amount of memory actually used by the application at any given time.  It is a trade-off everyone has to assess according to their own needs, I guess.

**Comment by interwq** [comment]
Update: [5.3 just released](https://github.com/jemalloc/jemalloc/releases/tag/5.3.0). Let's give it some time to settle, and then we will discuss about FreeBSD integration.

**Comment by jm-mikkelsen** [comment]
> @kempniu : forgot to mention, another option is to enable background threads (adding `background_thread:true` in `MALLOC_CONF`). It allows jemalloc to create its own threads which handles the purging and takes that load off the application threads. jemalloc can't really enable it by default but in the vast majority of use cases, it improves tail latency and smooths out the purging behavior (e.g. avoids RSS spikes when application threads go idle).

Unfortunately, the FreeBSD build of jemalloc seems to disable the background_thread option.

For example, on a FreeBSD 13.1 system:

$ MALLOC_CONF=background_thread:true sync
&lt;jemalloc>: option background_thread currently supports pthread only

**Comment by AdamPelc** [comment]
> Update: [5.3 just released](https://github.com/jemalloc/jemalloc/releases/tag/5.3.0). Let's give it some time to settle, and then we will discuss about FreeBSD integration.

Hi, do I understand that this problem was fixed in the 5.3 release? So this issue is still open just to keep track of the FreeBSD integration?
